File: 9лаб_16варКЗП/лаб9/Dictaphone.py
from AudioPlayer import AudioPlayer

from Speaker import Speaker
from BluetoothModule import BluetoothModule
from Battery import Battery

# Клас Dictaphone розширює функціональність AudioPlayer, додаючи можливість запису аудіо.
class Dictaphone(AudioPlayer):
    def __init__(self, speaker=None, bluetooth_module=None, battery=None):
        # Викликає конструктор батьківського класу AudioPlayer для ініціалізації динаміка, Bluetooth-модуля та батареї.
        super().__init__(speaker, bluetooth_module, battery)
        self.is_recording = False  # Змінна для відстеження стану запису (пише або не пише).

    # Метод для початку запису.
    def start_recording(self):
        # Перевіряє рівень заряду батареї, якщо батарея розряджена, запис не починається.
        if self.get_capacity() <= 0:
            print("Battery is empty. Cannot start recording.")
            return
        # Якщо запис ще не розпочато, починає запис і зменшує заряд батареї на 20 мА·год.
        if not self.is_recording:
            self.is_recording = True
            print("Recording started...")
            self.set_capacity(self.get_capacity() - 20)
        else:
            # Якщо запис вже триває, виводить повідомлення.
            print("Already recording.")

    # Метод для зупинки запису.
    def stop_recording(self):
        # Якщо диктофон записує, зупиняє запис.
        if self.is_recording:
            self.is_recording = False
            print("Recording stopped.")
        else:
            # Якщо запис не відбувається, виводить повідомлення.
            print("Not recording.")

    # Метод для відтворення записаного аудіо.
    def play_recording(self):
        # Перевіряє рівень заряду батареї, якщо батарея розряджена, відтворення неможливе.
        if self.get_capacity() <= 0:
            print("Battery is empty. Cannot play recording.")
            return
        # Відтворює запис і зменшує заряд батареї на 30 мА·год.
        print("Playing recorded audio...")
        self.set_capacity(self.get_capacity() - 30)

    # Перевизначений метод для виведення інформації про готовність диктофона до роботи.
    def device_functionality(self):
        print("Dictaphone is ready for recording and playback.")

    # Метод для формування рядкового представлення об'єкта Dictaphone.
    def __str__(self):
        # Повертає інформацію про стан запису та рівень заряду батареї.
        return f"Dictaphone{{isRecording={self.is_recording}, battery={self.get_capacity()}mAh}}"

File: 3лаб_16варКЗП/лаб3/Lab3/src/KI/Milian/Lab3/Dictaphone.java
package KI.Milian.Lab3;

import java.io.IOException;

/**
 * Інтерфейс для пристроїв, що можуть записувати звук.
 */
interface Recordable {
    /**
     * Метод для початку запису.
     */
    void startRecording();

    /**
     * Метод для зупинки запису.
     */
    void stopRecording();
}

/**
 * Клас, що описує диктофон. Розширює абстрактний клас AudioPlayer та реалізує інтерфейс Recordable.
 */
public class Dictaphone extends AudioPlayer implements Recordable {
    private boolean isRecording;

    public Dictaphone(Speaker speaker, BluetoothModule bluetoothModule, Battery battery) throws IOException {
        super(speaker, bluetoothModule, battery);
        this.isRecording = false;
    }

    public Dictaphone() throws IOException {
        super();
        this.isRecording = false;
    }

    @Override
    public void startRecording() {
        if (getCapacity() <= 0) {
            System.out.println("Battery is empty. Cannot start recording.");
            return;
        }
        if (!isRecording) {
            isRecording = true;
            System.out.println("Recording started...");
            setCapacity(getCapacity()- 20); // Витрачаємо 20 мАг на початок запису
        } else {
            System.out.println("Already recording.");
        }
    }

    @Override
    public void stopRecording() {
        if (isRecording) {
            isRecording = false;
            System.out.println("Recording stopped.");
        } else {
            System.out.println("Not recording.");
        }
    }

    @Override
    public void deviceFunctionality() throws IOException {
        System.out.println("Dictaphone is ready for recording and playback.");
        log("Dictaphone initialized.");
    }

    public void playRecording() throws IOException {
        if (getCapacity() <= 0) {
            System.out.println("Battery is empty. Cannot play recording.");
            return;
        }
        System.out.println("Playing recorded audio...");
        setCapacity(getCapacity()- 30); // Витрачаємо 30 мАг на відтворення запису
    }

    @Override
    public String toString() {
        return "Dictaphone{isRecording=" + isRecording + ", battery=" + getCapacity() + "mAh}";
    }
}
